esrally/utils/net.py

```diff
@@ -94,17 +95,64 @@ def __call__(self, bytes_amount):
+
+    return functools.reduce(urllib.parse.urljoin, [
+        "https://storage.googleapis.com/storage/v1/b/",
+        f"{quote(bucket_name.strip('/'), safe='')}/",
+        "o/",
+        f"{quote(bucket_path.strip('/'), safe='')}",
+        "?alt=media"
+    ])
+
+
+def _download_from_gcs_bucket(bucket_name, bucket_path, local_path, expected_size_in_bytes=None, progress_indicator=None):
+    # lazily initialize Google Cloud Storage support - we might not need it
+    import google.oauth2.credentials
+    import google.auth.transport.requests as tr_requests
+    import google.auth
+    # Using Google Resumable Media as the standard storage library doesn't support progress
+    # (https://github.com/googleapis/python-storage/issues/27)
+    from google.resumable_media.requests import ChunkedDownload
+    ro_scope = "https://www.googleapis.com/auth/devstorage.read_only"
+
+    access_token = os.environ.get("GOOGLE_AUTH_TOKEN")
+    if access_token:
+        credentials = google.oauth2.credentials.Credentials(token=access_token, scopes=(ro_scope, ))
+    else:
+        # https://google-auth.readthedocs.io/en/latest/user-guide.html
+        credentials, _ = google.auth.default(scopes=(ro_scope,))
+
+    transport = tr_requests.AuthorizedSession(credentials)
+    chunk_size = 50 * 1024 * 1024  # 50MB
+
+    with open(local_path, "wb") as local_fp:
+        media_url = _build_gcs_object_url(bucket_name, bucket_path)
+        download = ChunkedDownload(media_url, chunk_size, local_fp)
+        # allow us to calculate the total bytes
+        download.consume_next_chunk(transport)
+        if not expected_size_in_bytes:
+            expected_size_in_bytes = download.total_bytes
+        while not download.finished:
+            if progress_indicator and download.bytes_downloaded and download.total_bytes:
+                progress_indicator(download.bytes_downloaded, expected_size_in_bytes)
+            download.consume_next_chunk(transport)
+
+
+def download_from_bucket(blobstore, url, local_path, expected_size_in_bytes=None, progress_indicator=None):
+    blob_downloader = {"s3": _download_from_s3_bucket, "gs": _download_from_gcs_bucket}
     logger = logging.getLogger(__name__)
 
-    bucket_and_path = url[5:]
+    bucket_and_path = url[5:]  # s3:// or gs:// prefix for now
     bucket_end_index = bucket_and_path.find("/")
     bucket = bucket_and_path[:bucket_end_index]
     # we need to remove the leading "/"
     bucket_path = bucket_and_path[bucket_end_index + 1:]
 
-    logger.info("Downloading from S3 bucket [%s] and path [%s] to [%s].", bucket, bucket_path, local_path)
-    _download_from_s3_bucket(bucket, bucket_path, local_path, expected_size_in_bytes, progress_indicator)
+    logger.info("Downloading from [%s] bucket [%s] and path [%s] to [%s].", blobstore, bucket, bucket_path, local_path)
+    blob_downloader[blobstore](bucket, bucket_path, local_path, expected_size_in_bytes, progress_indicator)
```

Comment:
entirely up to you, but i find `if` statements easier to reason about than looking up some string->function call magic

Reply:
Thanks, used it because it is a pattern we use elsewhere too.